### DISCUSSION TITLE
Multiple buddy box profiles

### DISCUSSION THREAD

**Discussion by sande005** [EdgeTX/edgetx#2704] [Ideas]
As an instructor using the traditional non-MPM trainer setup, using a single club issued trainer, I've occasionally had the situation where  there are two students, each with their own RX, to teach on the same day.  It would be nice to have 2 to 4 buddy box configurations saved in the radio, that can be easily swapped out depending on the specific slave TX being used.  Currently that means reconfiguring the trainer setups each time an alternate student wants to fly with their box.
This would require a named field for the profile, and a way to add/edit/delete same from the Trainer setup page.  On the order of "My Taranis", "Paul's Spektrum", "Sue's Futaba", each with their unique channel order, scaling, offset, reversing, etc. needed to match the Master.
Implementation possibilities:
1. Make the active profile Global, in the sense that it can be chosen from a System screen, and applies to all models that use the trainer function.
2. Make it active only for a specific model, so there would have to be copies of a model, only variable by the buddy box profile assigned.  ("Apprentice - Paul", "Apprentice - Sue", etc.)
3. Make it easily "choose-able" within a model.
I'll leave it to others to debate the preferred method, and the complications if warnings/confirmations needed to guard against accidentally using the wrong profile for flight.

**Comment by pfeerick** [comment]
To allow for all three (or really two options as I see it), could look at the "ADC Jitter filter" option... which can either be "global, on, off" on a per-model basis - so you both have a global setting, and you also have a per-model setting. Up to you how you use them. 

As far as warnings, well, this affects the student radio input, not the master radio control. So I think it's a little hard to say - perhaps an option for a message to be displayed on selecting the model to remind of the buddy box set? Which, of course, only protects if you change models, or power cycle the radio between students. i.e. is it a check that gives a false sense of security as it only works part of the time?

**Comment by ChrisOhara57** [comment]
Hi
I do also a lot of buddy flying. (with OpenTx)
If you design your Trainer settings in the **input lines of your model** you can do everything you request.
(do not use channel mixer lines)
I have 9 Modell with different settings on my Taranis 9xlite.
You can use different lines in inputs (it does work similiar to the mixing of Channels). 
So define two lines per input:  one for Trainer (with source Ail) and one for Student (with source TR1)  etc.
You can use different weight and Expo and Hardware switch or Logical switch  or just add up with different weight.
I even has two copies of the same model one as buddy and one as spotter.
I never change hardware settings in the basic Radio screen.

I've attached a description and screenshot of an example.

[Buddy Flying.pdf](https://github.com/EdgeTX/edgetx/files/10347820/Buddy.Flying.pdf)

**Comment by ChrisOhara57** [comment]
Sorry I have to correct what I said.
My first chart I uploaded does work.
But why inputs do have more than one line there seems to be no mixing.
It looks like you have to use switches and the system takes only the first one that is active.
So it is possible to use inputs alone if you want a complete switch from master to student and vice versa.
For overruling you need mixer.

Saying this I know to achieve the same functionality and I had already tested once.
It is a little bit more complicated
You take 4 channels you normaly do not use (Ch29, Ch30, Ch31, Ch32)
You always mix two sources (example: Stick of Aileron, and TR1 both noTrim).
Then you use the result as source for Input I01 (what is Aileron Source for Channel mixer). Do Also select TrmA as trim.

There is still the advantage that you don't have to touch any channel mixer.
And the definitions can be done by LUA because normally they are identical.
(or copied with companion)
And all option are stored on model level as required

![Screenshot 2023-01-06 002742](https://user-images.githubusercontent.com/80360397/210899484-5de89c10-eea8-4189-9951-0d7d86fac0cb.jpg)

**Comment by sande005** [comment]
Interesting ideas, but doesn't address the concept of rapid switching between two or more very different buddy boxes at all - one say a Futaba, and one a Spektrum, each with it's own unique channel order, ppmus min/max/center value, trim biases and reversing.

**Comment by ChrisOhara57** [comment]
That's not so easy to achieve.

Imagine you do no calibration and change of weight in the basic trainer settings of your hardware.
(setting your (already modified) radio to that state you have to use companion)
So your channel order center point of sticks  ....     are your TR1.... values as they come from the students radio.

You have to have a copy of the model for every student radio type.
Change the source in the mixer lines to the correct TR.. sources (Channel order)
Then you have to use offset and weight in the mixer line of TR1 to do the calibration.

This is possible but I guess I would also need about 15 minute time for every new configuration.

**Comment by ChrisOhara57** [comment]
I've done a concept that does fulfill all the requirements.
All information/definitions are now part of the modell-data.

With the script I have written I could also change each of my existing models for buddy flying in less than 3 minutes (without companion).
I've also written scripts that do support the channel-mapping and calibration of new student radios. (similar to simulators)

I uploaded a detailed description on Aloft OpenTX forum.

https://forum.alofthobbies.com/index.php?threads/trainer-student-buddy-flying-advanced-settings-in-opentx.3278/

https://openrcforums.com/forum/viewtopic.php?t=14063&sid=df6d3f7f597ff57eb767719a8d20ae28
